client/client.go

```diff
@@ -494,26 +498,55 @@ func (c *Client) CmdChallenge(status chan *CmdStatus, args ...string) {
 	lastBlock := atomic.LoadUint64(&c.lastBlock)
 	if c.params.IsChallengeResponsePhase(lastBlock + 1) {
 		status <- &CmdStatus{Msg: "Waiting for next epoch"}
-		next := c.params.DepositDoneBlock(c.params.ExitEpoch(lastBlock + 1))
+		next := c.params.DepositDoneBlock(c.params.DepositEpoch(lastBlock + 1))
 		if err := c.ethClient.WaitForBlock(c.Ctx(), next); err != nil {
 			status <- &CmdStatus{Err: err}
 			return
 		}
 	}
 
+	// Get the BP of the sealed epoch.
+	lastBlock = atomic.LoadUint64(&c.lastBlock) + 1
+	sealed := c.params.SealedEpoch(lastBlock)
+	c.balMtx.Lock()
+	bal, ok := c.balances[sealed]
+	c.balMtx.Unlock()
+	if !ok || bal.Bal == nil {
+		status <- &CmdStatus{Err: fmt.Errorf("No Balance proof available")}
+		return
+	}
+
+	c.challenge(status, bal.Bal)
+}
+
+func (c *Client) challengeDeposit(status chan *CmdStatus) {
+	tx, err := c.sendTx("ChallengeDeposit", func(auth *bind.TransactOpts) (*types.Transaction, error) {
+		return c.contract.ChallengeDeposit(auth)
+	}, status)
+	if err != nil {
+		status <- &CmdStatus{Err: fmt.Errorf("ChallengeDeposit TX: %w", err)}
+		return
+	}
+	c.waitForReponse(status, tx.BlockNumber.Uint64())
+}
+
+func (c *Client) challenge(status chan *CmdStatus, bp *tee.BalanceProof) {
 	tx, err := c.sendTx("Challenge", func(auth *bind.TransactOpts) (*types.Transaction, error) {
-		return c.contract.Challenge(auth)
+		return c.contract.Challenge(auth, bp.Balance.ToEthBals(), bp.Sig)
 	}, status)
 	if err != nil {
 		status <- &CmdStatus{Err: fmt.Errorf("Challenge TX: %w", err)}
 		return
 	}
+	c.waitForReponse(status, tx.BlockNumber.Uint64())
+}
 
+func (c *Client) waitForReponse(status chan *CmdStatus, block uint64) {
```

Comment:
```suggestion
func (c *Client) waitForResponseOrWithdraw(status chan *CmdStatus, block uint64) {
```
there was a missing `s` in `Response` and since this function is actually also doing something (withdrawing), I'd state this explicitly in its name.

Reply:
\+ added a comment